src/client/network/nusight_network.ts

```diff
@@ -6,8 +6,69 @@ import { WebSocketProxyNUClearNetClient } from '../nuclearnet/web_socket_proxy_n
 import { MessageTypePath } from './message_type_names'
 import { RobotModel } from '../components/robot/model'
 import { AppModel } from '../components/app/model'
+import * as DecodeWorker from './decode.worker.ts'
 
-const HEADER_SIZE = 9
+export class MessageDecoder {
+  public constructor(private free: Worker[],
+                     private busy: Worker[],
+                     private callbacks: {[key: string]: (message: any) => void},
+                     private tokenSource : number) {
+
+    // Register all our completion handlers
+    free.forEach((worker) => {
+      worker.addEventListener('message', (ev: MessageEvent) => {
+        this.onTaskComplete(worker, ev.data)
+      })
+    })
+  }
+
+  public static of() {
+    const free = []
+
+    // We make a webworker for each cpu core we have
+    for (let i = 0; i < navigator.hardwareConcurrency; ++i) {
+      free.push(new DecodeWorker())
+    }
+    return new MessageDecoder(free, [], {}, 0)
+  }
+
+  public decode<T>(type: string, packet: NUClearNetPacket, cb: (message: T) => void) {
+
+    // Try to get a worker to process with
+    let worker = this.free.pop()
+
+    // If we don't have a free worker but we are reliable, get a busy one
+    worker = worker === undefined && packet.reliable ? this.busy.shift() : worker
+
+    // We drop packets if they are unreliable and we are too busy to process them
+    if (worker !== undefined) {
+      // Store our callback with a unique token so we can find it later
+      const token = ++this.tokenSource
+      this.callbacks[token] = cb
+
+      // Execute on the webworker
+      worker.postMessage({
+        'type': type,
+        'token': token,
+        'payload': packet.payload,
+      }, [packet.payload])
+      this.busy.push(worker)
+    }
+  }
+
+  private onTaskComplete(worker: Worker, msg: {token: string, output: any}) {
+
+    // Find and execute our callback
+    this.callbacks[msg.token](msg.output)
+    delete this.callbacks[msg.token]
+
+    // If this worker has finished all its tasks move it to the free list
+    const w = this.busy.splice(this.busy.indexOf(worker), 1)[0]
```

Comment:
This is not ideal, i'd like to only remove it from busy if it actually has no other task it's running